pkg/hook/register.go

```diff
@@ -66,4 +66,49 @@ func RegisterDefaultHooks() {
 	RegisterMethodHook("ExecuteTemplate", "*html/template.Template", "HtmlTemplateExecuteTemplate")
 	RegisterMethodHook("Execute", "*text/template.Template", "TextTemplateExecute")
 	RegisterMethodHook("ExecuteTemplate", "*text/template.Template", "TextTemplateExecuteTemplate")
+
+	// Hooks for path traversal bugs
+	RegisterFunctionHook("Chdir", "os", "OsChdir")
+	RegisterFunctionHook("Chmod", "os", "OsChmod")
+	RegisterFunctionHook("Chown", "os", "OsChown")
+	RegisterFunctionHook("Chtimes", "os", "OsChtimes")
+	RegisterFunctionHook("Create", "os", "OsCreate")
+	RegisterFunctionHook("CreateTemp", "os", "OsCreateTemp")
+	RegisterFunctionHook("DirFS", "os", "OsDirFS")
+	RegisterFunctionHook("Lchown", "os", "OsLchown")
+	RegisterFunctionHook("Link", "os", "OsLink")
+	RegisterFunctionHook("Lstat", "os", "OsLstat")
+	RegisterFunctionHook("Mkdir", "os", "OsMkdir")
+	RegisterFunctionHook("MkdirAll", "os", "OsMkdirAll")
+	RegisterFunctionHook("MkdirTemp", "os", "OsMkdirTemp")
+	RegisterFunctionHook("NewFile", "os", "OsNewFile")
+	RegisterFunctionHook("Open", "os", "OsOpen")
+	RegisterFunctionHook("OpenFile", "os", "OsOpenFile")
+	RegisterFunctionHook("ReadDir", "os", "OsReadDir")
+	RegisterFunctionHook("ReadFile", "os", "OsReadFile")
+	RegisterFunctionHook("Readlink", "os", "OsReadlink")
+	RegisterFunctionHook("Remove", "os", "OsRemove")
+	RegisterFunctionHook("RemoveAll", "os", "OsRemoveAll")
+	RegisterFunctionHook("Rename", "os", "OsRename")
+	RegisterFunctionHook("Stat", "os", "OsStat")
+	RegisterFunctionHook("Symlink", "os", "OsSymlink")
+	RegisterFunctionHook("Truncate", "os", "OsTruncate")
+	RegisterFunctionHook("WriteFile", "os", "OsWriteFile")
+
+	RegisterFunctionHook("ReadDir", "io/ioutil", "IoutilReadDir")
+	RegisterFunctionHook("ReadFile", "io/ioutil", "IoutilReadFile")
+	RegisterFunctionHook("TempDir", "io/ioutil", "IoutilTempDir")
+	RegisterFunctionHook("TempFile", "io/ioutil", "IoutilTempFile")
+	RegisterFunctionHook("WriteFile", "io/ioutil", "IoutilWriteFile")
+
+	RegisterFunctionHook("Walk", "path/filepath", "FilepathWalk")
+	RegisterFunctionHook("WalkDir", "path/filepath", "FilepathWalkDir")
+
+	RegisterFunctionHook("FileInfoToDirEntry", "io/fs", "FsFileInfoToDirEntry")
```

Comment:
Same question as above: Is this problematic? If so we should also intercept `Glob` and other functions returning file listings, if not remove the intercepts.

Reply:
IMO, I agree that these candidates do not strictly help doing a *path traversal*. As they still could be user-controlled and leak information they're still reasonable hooking spots though. So, I figured we could just get them as a bonus here. Then again, others may flag these as *false positives* for the same reason.

If we want to be explicit, we likely should separate them in something like a "information leak detector", whereas we then would need to specify strict borders in what counts as what so we do **not** have any overlap between sanitizers and have misleading error reports.

Reply:
The scenario I have in mind here is when these functions are used as the first step in a path traversal attack where you can control which directory is traversed. The traversal code may then read or write some files. At that point, it might be too late for the detector as we might not have the `../../xxx` in the paths provided as arguments there.

```golang
func visit(path string, di fs.DirEntry, err error) error {
    abs, err := filepath.Abs(path)
    if err != nil {
      return err
    }
    // At this point, the file APIs reading from or writing to files will get the absolute
    // path and our hooks will not detect the path traversal
    readOrWriteFile(abs)
    return nil
}

func traverseDir(dir string) error {
    return filepath.WalkDir(root, visit)
}
```

Reply:
That was also my understanding, thx for clarification.